banyan/banyan/sessions.py

```diff
@@ -1,8 +1,194 @@
+    return _s3_client.get_object(
+        Bucket="banyan-executor",
+        Key="executor.py"
+    )
+
+
+def _get_executor_lambda_zip(site_packages_dir: str):
+    # Construct zipfile with site_packages_dir if provided
+    zipfile_name = "executor_lambda_code"
+    if site_packages_dir:
+        shutil.make_archive(zipfile_name, "zip", root_dir=site_packages_dir, base_dir=".")
+    else:
+        # Create empty zipfile
+        with zipfile.ZipFile(zipfile_name + ".zip", "w") as f:
+            pass
+    # Add executor code file
+    executor_code = _get_executor_code_from_s3()["Body"].read().decode("utf-8")
+    with zipfile.ZipFile(zipfile_name + ".zip", "a") as f:
+        f.writestr("lambda_function.py", executor_code)
+    # Upload to S3. First create a S3 bucket if not already existing
+    # and then upload the zip file.
+    bucket_name = "banyan-assets"
+    s3_key = zipfile_name + ".zip"
+    try:
+        _s3_client.create_bucket(
+            Bucket=bucket_name,
+            CreateBucketConfiguration={"LocationConstraint": _s3_client.meta.region_name}
+        )
+        _s3_client.get_waiter("bucket_exists").wait(Bucket=bucket_name)
+    except ClientError as error:
+        if not error.response["Error"]["Code"] in ["BucketAlreadyExists", "BucketAlreadyOwnedByYou"]:
+            raise
+    with open(zipfile_name + ".zip", "rb") as f:
+        _s3_client.put_object(
+            Bucket=bucket_name,
+            Body=f,
+            Key=s3_key
+        )
+    return bucket_name, s3_key
+
+
+def _update_executor_code(
+    executor_lambda_function_name: str,
+    lambda_last_modified: int
+):
+    executor_code_object =_get_executor_code_from_s3()
+    executor_code_last_modified = executor_code_object["LastModified"].timestamp()
+    if executor_code_last_modified > lambda_last_modified:
+        # Update executor code
+        bucket_name, s3_key = _get_executor_lambda_zip()
+        _s3_client.update_function_code(
+            FunctionName=executor_lambda_function_name,
+            S3Bucket=bucket_name,
+            S3Key=s3_key,
+        )
+
+
+def _create_executor_lambda_iam_role():
+    assume_role_policy = {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Effect": "Allow",
+                "Principal": {
+                    "Service": "lambda.amazonaws.com"
+                },
+                "Action": "sts:AssumeRole"
+            },
+        ]
+    }
+    basic_lambda_policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+    iam_role_name = "banyan-executor-lambda-role"
+    try:
+        role = _iam_client.create_role(
+            RoleName=iam_role_name,
+            AssumeRolePolicyDocument=json.dumps(assume_role_policy)
+        )
+        _iam_client.get_waiter("role_exists").wait(RoleName=iam_role_name)
+        _iam_client.attach_role_policy(RoleName=iam_role_name, PolicyArn=basic_lambda_policy_arn)
+        time.sleep(5)
```

Comment:
Is this 5 second thing an artifact from debugging or was it critical for this to work? Is it critical every time or just the first time a user uses Banyan? and if it takes longer than 5s what will happen?

Doing a while loop to ensure that a resource exists shouldn't be hard. We did this for SQS queues in several places.

Reply:
That was intentional; I will add a comment.

I think it usually takes 5-10 seconds to reflect, but a try loop would be better to add. I didn't find an easy way to check this so just added a sleep for now.